### PR TITLE
Update docs before `0.19.0` release

### PR DESCRIPTION
**Pull request by ddoktorski**
## Introduced changes

- Update docs before `0.19.0` release

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: docs/migration_guide.rst [hunk 1]
```diff
@@ -9,32 +9,44 @@ Version 0.19.0 of **starknet.py** comes with support for RPC 0.6.0!
...
 
 .. currentmodule:: starknet_py.net.account.account
 
+- :meth:`~Account.execute` has been renamed to :meth:`~Account.execute_v1`
 - :meth:`~Account.execute_v3` has been added.
-- :meth:`~Account.sign_declare_v3_transaction`, :meth:`~Account.sign_deploy_account_v3_transaction` and :meth:`~Account.sign_invoke_v3_transaction` have been added.
-- :meth:`~Account.sign_declare_transaction`, :meth:`~Account.sign_deploy_account_transaction` and :meth:`~Account.sign_invoke_transaction` have been renamed to :meth:`~Account.sign_declare_v1_transaction`, :meth:`~Account.sign_deploy_account_v1_transaction` and :meth:`~Account.sign_invoke_v1_transaction` respectively.
+- :meth:`~Account.deploy_account` has been renamed to :meth:`~Account.deploy_account_v1`
+- :meth:`~Account.deploy_account_v3` has been added
+- :meth:`~Account.sign_declare_v3_transaction`, :meth:`~Account.sign_deploy_account_v3_transaction` and :meth:`~Account.sign_invoke_v3_transaction` have been added
```
Comment (ddoktorski) [inline docs/migration_guide.rst:23]: Given that we already have all these breaking changes, what do you think about removing the word `transaction` from the `Account` methods? This would result in names like `sign_invoke_v1`, `sign_declare_v3`, etc.

File: docs/guide/deploying_contracts.rst [hunk 2]
```diff
@@ -4,27 +4,20 @@ Deploying contracts
 Declaring contracts
 -------------------
 
-Since Cairo 0.10.0 Declare transactions can be signed and in the future, declaring without signing the transaction
-(and without paying the fee) will be impossible. That is why :ref:`Account` has
-:meth:`sign_declare_transaction()` method.
+Declare transaction can be sent as either version 1 or 3. To sign a declare transaction, please use :meth:`sign_declare_transaction_v1()` or :meth:`sign_declare_transaction_v3()` method.
```
Comment (THenry14) [inline docs/guide/deploying_contracts.rst:7]:
```suggestion
A declare transaction can be issued in either version 1 or version 3. To sign a declare transaction, you should utilize the sign_declare_transaction_v1() or sign_declare_transaction_v3() method, respectively.
```
maybe?

File: docs/guide/using_existing_contracts.rst [hunk 3]
```diff
@@ -37,47 +37,47 @@ If you do not have ABI statically, but you know the interface of the contract on
...
+when preparing function calls with ``.prepare_invoke_v1()`` or ``.prepare_invoke_v3()``.
 
 .. code-block:: python
 
-    await contract.functions["put"].invoke(k, v, max_fee=5000)
+    await contract.functions["put"].invoke_v1(k, v, max_fee=5000)
 
-When max_fee is specified when preparing a call, you can invoke it without
-``max_fee``.
+When max_fee is specified when preparing a call, you can invoke it without ``max_fee``. It works similarly for ``l1_resource_bounds`` and ``prepare_invoke_v3()``.
```
Comment (THenry14) [inline docs/guide/using_existing_contracts.rst:48]: I don't quite understand it to be honest. what does `when max_fee is specified (...) you can invoke it without max_fee` mean?

File: docs/guide/using_existing_contracts.rst [hunk 4]
```diff
@@ -37,47 +37,47 @@ If you do not have ABI statically, but you know the interface of the contract on
...
-    prepared_call = contract.function["put"].prepare(k, v, max_fee=5000)
+    prepared_call = contract.function["put"].prepare_invoke_v1(k, v, max_fee=5000)
     await prepared_call.invoke()
 
 .. warning::
 
-    If ``max_fee`` is not specified at any step it will default to ``None``,
-    and will raise an exception when invoking a transaction.
+    For V1 transactions if ``max_fee`` is not specified at any step it will default to ``None``,
+    and will raise an exception when invoking a transaction. The same applies to ``l1_resource_bounds`` and V3 transactions.
```
Comment (THenry14) [inline docs/guide/using_existing_contracts.rst:58]: shouldn't we auto-estimate? how does this work, is auto-estimation off by default? or is it exclusive to explicitly setting max_fee to anything?

File: docs/migration_guide.rst (hunk 1 above)
Comment (THenry14) [reply docs/migration_guide.rst:23]: I'm fine with that

File: docs/migration_guide.rst [hunk 5]
```diff
@@ -9,32 +9,44 @@ Version 0.19.0 of **starknet.py** comes with support for RPC 0.6.0!
...
+- :meth:`~Account.sign_declare_transaction`, :meth:`~Account.sign_deploy_account_transaction` and :meth:`~Account.sign_invoke_transaction` have been renamed to :meth:`~Account.sign_declare_v1_transaction`, :meth:`~Account.sign_deploy_account_v1_transaction` and :meth:`~Account.sign_invoke_v1_transaction` respectively
 
 All new functions with ``v3`` in their name operate similarly to their ``v1`` and ``v2`` counterparts.
 Unlike their ``v1`` counterparts, ``v3`` transaction fees are paid in Fri (10^-18 STRK). Therefore,  ``max_fee`` parameter, which is typically set in Wei, is not applicable for ``v3`` functions. Instead, ``l1_resource_bounds`` parameter is utilized to limit the Fri amount used.
```
Comment (THenry14) [inline docs/migration_guide.rst:27]:
```suggestion
Unlike their ``v1`` counterparts however, ``v3`` transaction fees are paid in Fri (10^-18 STRK). Therefore,  ``max_fee`` parameter, which is typically set in Wei, is not applicable for ``v3`` functions. Instead, ``l1_resource_bounds`` parameter is utilized to limit the Fri amount used.
```

**Comment by THenry14** [comment]
Let's also improve estimate fee section/guides as requested by SW on slack channel

File: docs/guide/using_existing_contracts.rst (hunk 4 above)
Comment (ddoktorski) [reply docs/guide/using_existing_contracts.rst:58]: By default, the auto estimation is not enabled. The user must explicitly turn it on by setting `auto_estimate=True`. However, using both `auto_estimate` and `max_fee `simultaneously is not allowed and doing so will cause an error.

I think we shouldn't use auto estimation as a default when `max_fee` is not provided, since it could lead to high and unexpected costs for users. We point out in the documentation that auto estimation is a feature for convenience and should be used mostly for testing.

Comment (THenry14) [reply docs/guide/using_existing_contracts.rst:58]: > By default, the auto estimation is not enabled. The user must explicitly turn it on by setting auto_estimate=True. However, using both auto_estimate and max_fee simultaneously is not allowed and doing so will cause an error.

Sure, makes sense

> I think we shouldn't use auto estimation as a default when max_fee is not provided, since it could lead to high and unexpected costs for users. We point out in the documentation that auto estimation is a feature for convenience and should be used mostly for testing.

That's fine as well, but let's make this info more prominent - it should be more discoverable that auto estimation is possible

File: docs/guide/using_existing_contracts.rst (hunk 3 above)
Comment (ddoktorski) [reply docs/guide/using_existing_contracts.rst:48]: When `max_fee` is specified in either `prepare_invoke_v1` or `prepare_invoke_v3`, the prepared call can subsequently be invoked without redefining `max_fee`, simply by using `prepared_call.invoke()`, rather than `prepared_call.invoke(max_fee=...)`. I agree the wording isn't optimal, I will try to rephrase it.

File: docs/guide/account_and_client.rst [hunk 6]
```diff
@@ -4,16 +4,42 @@ Account and Client
...
     :language: python
     :dedent: 4
 
+Transaction Fee
+---------------
+
+All methods within the :ref:`Account` that involve on-chain modifications require either to specify a maximum transaction fee or to use auto estimation.
+In the case of V1 and V2 transactions, the transaction fee, denoted in Wei, is configured by the ``max_fee`` parameter.
+For V3 transactions, however, the fee is expressed in Fri and is determined by the ``l1_resource_bounds`` parameter.
+To enable auto estimation, assign ``True`` to the ``auto_estimate`` parameter.
```
Comment (THenry14) [inline docs/guide/account_and_client.rst:20]:
```suggestion
To enable auto estimation, set the ``auto_estimate`` parameter to `True``.
```

File: docs/guide/account_and_client.rst [hunk 7]
```diff
@@ -4,16 +4,42 @@ Account and Client
...
+
+.. code-block:: python
+
+    resp = await account.execute_v1(calls=call, auto_estimate=True)
+
+.. warning::
+
+    Do not use automatic fee estimation in production code! It may lead to
+    very high fees paid as the amount returned by :meth:`~starknet_py.net.full_node_client.FullNodeClient.estimate_fee` may be arbitrarily large.
```
Comment (THenry14) [inline docs/guide/account_and_client.rst:29]: Let's rephrase this, to something like "it is discouraged to use auto estimation on prod"
if someone wants to use it, and is confident enough, let them.

File: docs/guide/account_and_client.rst [hunk 8]
```diff
@@ -4,16 +4,42 @@ Account and Client
...
+
+The returned estimated fee is multiplied by ``1.5`` for V1 and V2 transactions.
```
Comment (THenry14) [inline docs/guide/account_and_client.rst:31]:
```suggestion
The returned estimated fee is multiplied by ``1.5`` for V1 and V2 transactions to mitigate fluctuations in price.
```
maybe? I guess more info about this decision won't harm anyone

File: docs/guide/account_and_client.rst [hunk 9]
```diff
@@ -4,16 +4,42 @@ Account and Client
...
+For V3 transactions, ``max_amount`` and ``max_price_per_unit`` are scaled by ``1.1`` and ``1.5``.
```
Comment (THenry14) [inline docs/guide/account_and_client.rst:32]:
```suggestion
For V3 transactions, ``max_amount`` and ``max_price_per_unit`` are scaled by ``1.1`` and ``1.5`` respectively.
```

**Review by THenry14 — APPROVED** [review]

File: docs/guide/account_and_client.rst [hunk 10]
```diff
@@ -4,16 +4,41 @@ Account and Client
...
+These methods will send :class:`~starknet_py.net.models.InvokeV1` and :class:`~starknet_py.net.models.InvokeV3` transactions respectively. To read about differences between transaction versions please visit `transaction types <https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/transactions>`_ from the Starknet docs.
 
 .. codesnippet:: ../../starknet_py/tests/e2e/docs/guide/test_executing_transactions.py
     :language: python
     :dedent: 4
 
+Transaction Fee
+---------------
+
+All methods within the :ref:`Account` that involve on-chain modifications require either to specify a maximum transaction fee or to use auto estimation.
```
Comment (DelevoXDG) [inline docs/guide/account_and_client.rst:17]: Nit:
```suggestion
All methods within the :ref:`Account` that involve on-chain modifications require either specifying a maximum transaction fee or using auto estimation.
```

File: docs/guide/account_and_client.rst [hunk 11]
```diff
@@ -4,16 +4,41 @@ Account and Client
...
+For V3 transactions, however, the fee is expressed in Fri and is determined by the ``l1_resource_bounds`` parameter.
+To enable auto estimation, set the ``auto_estimate`` parameter to ``True``.
+
+.. code-block:: python
+
+    resp = await account.execute_v1(calls=call, auto_estimate=True)
+
+.. warning::
+
+    It is strongly discouraged to use automatic fee estimation in production code as it may lead to unexpectedly high fee.
```
Comment (DelevoXDG) [inline docs/guide/account_and_client.rst:28]: Nit:
```suggestion
    It is strongly discouraged to use automatic fee estimation in production code as it may lead to an unexpectedly high fee.
```

File: docs/guide/deploying_contracts.rst [hunk 12]
```diff
@@ -4,27 +4,21 @@ Deploying contracts
 Declaring contracts
 -------------------
 
-Since Cairo 0.10.0 Declare transactions can be signed and in the future, declaring without signing the transaction
-(and without paying the fee) will be impossible. That is why :ref:`Account` has
-:meth:`sign_declare_transaction()` method.
+A declare transaction can be issued in either version 1 or version 3.
```
Comment (DelevoXDG) [inline docs/guide/deploying_contracts.rst:7]: What about declare v2?

File: docs/guide/deploying_contracts.rst [hunk 13]
```diff
@@ -75,9 +68,7 @@ Cairo1 contracts
 Declaring Cairo1 contracts
 ##########################
 
-| Starknet 0.11 introduced the ability to declare contracts written in Cairo1!
-
-To declare a new contract, Declare v2 or Declare v3 transaction has to be sent.
+To declare a contract in Cairo version 1 or higher, Declare V2 or Declare V3 transaction has to be sent.
```
Comment (DelevoXDG) [inline docs/guide/deploying_contracts.rst:71]: Nit: maybe also add a similar note above for Cairo 0?

File: docs/guide/using_existing_contracts.rst (hunk 4 above)
Comment (DelevoXDG) [reply docs/guide/using_existing_contracts.rst:62]: I'm not sure it was addressed, I suggest 
```suggestion
    For V1 transactions if ``max_fee`` is not specified at any step it will default to ``None``,
    and will raise an exception when invoking a transaction, unless `auto_estimate` is specified and is set to `True`. The same applies to ``l1_resource_bounds`` and V3 transactions.
```

File: docs/guide/using_existing_contracts.rst [hunk 14]
```diff
@@ -37,77 +37,79 @@ If you do not have ABI statically, but you know the interface of the contract on
 Fees
 ----
 
-starknet.py requires you to specify amount of Wei you
-are willing to pay either when making ``.invoke()`` transactions or when preparing
-function calls with ``.prepare()``.
+.. currentmodule:: starknet_py.contract
+
+Starknet.py requires you to specify amount of Wei (for V1 transaction) or Fri (for V3 transaction) you
+are willing to pay when executing either :meth:`~ContractFunction.invoke_v1` or :meth:`~ContractFunction.invoke_v3` transactions.
```
Comment (DelevoXDG) [inline docs/guide/using_existing_contracts.rst:43]: How about we mention right away at the top that `"Alternatively, you can estimate fee automatically by..."`

File: docs/migration_guide.rst [hunk 15]
```diff
@@ -9,32 +9,44 @@ Version 0.19.0 of **starknet.py** comes with support for RPC 0.6.0!
 
 .. currentmodule:: starknet_py.net.client_models
 
-New classes added to mirror the recent changes in the RPC v0.6.0 specification include:
+New classes added to mirror the recent changes in the RPC v0.6.0 specification that include:
```
Comment (DelevoXDG) [inline docs/migration_guide.rst:12]: Why this change? 🤭

File: docs/account_creation.rst [hunk 16]
```diff
@@ -32,6 +32,6 @@ Here is step by step example:
 
     If you are experiencing transaction failures with ``FEE_TRANSFER_FAILURE``
     make sure that the address you are trying to deploy is prefunded with enough
-    tokens, and verify that ``max_fee`` argument
-    in :meth:`~starknet_py.net.account.account.Account.sign_deploy_account_transaction` is set
+    tokens, and verify that ``max_fee`` argument in :meth:`~starknet_py.net.account.account.Account.sign_deploy_account_v1_transaction`
+    or ``l1_resource_bounds`` argument in :meth:`~starknet_py.net.account.account.Account.sign_deploy_account_v3_transaction` is set
```
[the remark this answers is not in the document]
Comment (ddoktorski) [reply docs/account_creation.rst]: Thanks! Should be fixed now.

File: docs/guide/deploying_contracts.rst (hunk 12 above)
Comment (ddoktorski) [reply docs/guide/deploying_contracts.rst]: Added info also about v2.

File: docs/guide/deploying_contracts.rst (hunk 13 above)
Comment (ddoktorski) [reply docs/guide/deploying_contracts.rst:71]: Added info about it under `Declaring contracts` section.

File: docs/guide/using_existing_contracts.rst (hunk 14 above)
Comment (ddoktorski) [reply docs/guide/using_existing_contracts.rst:43]: I'd prefer to keep the current version, as incorporating all the information about fee estimation here would overly extend the `Fees` section.

Comment (DelevoXDG) [reply docs/guide/using_existing_contracts.rst:43]: Current wording with `requires` implies there's no way to send an invoke transaction without setting max fee. The intention is not to provide all the information, but merely show the user know this is possible.

**Review by DelevoXDG — COMMENTED** [review]
Left two comments, please re-request when you reply / address those

File: docs/guide/using_existing_contracts.rst (hunk 14 above)
Comment (ddoktorski) [reply docs/guide/using_existing_contracts.rst:43]: Does it make sense to add `Alternatively, you can estimate fee automatically, as described in the 'Automatic fee estimation' section below.`?

Comment (DelevoXDG) [reply docs/guide/using_existing_contracts.rst:43]: Sounds good 👍🏻

**Review by DelevoXDG — APPROVED** [review]
LGTM

image
